datumaro/plugins/data_formats/datumaro_binary/base.py

```diff
@@ -91,27 +104,75 @@ def _read_media_type(self):
         else:
             raise NotImplementedError(f"media_type={media_type} is currently not supported.")
 
-    def _read_items(self):
+    def _read_items(self) -> None:
         (n_blob_sizes_bytes,) = struct.unpack("<I", self._fp.read(4))
         blob_sizes_bytes = self._crypter.decrypt(self._fp.read(n_blob_sizes_bytes))
         blob_sizes, _ = IntListMapper.backward(blob_sizes_bytes, 0)
 
-        self._items = []
-
         media_path_prefix = {
             MediaType.IMAGE: osp.join(self._images_dir, self._subset),
             MediaType.POINT_CLOUD: osp.join(self._pcd_dir, self._subset),
         }
 
-        # For each blob, we decrypt the blob first, then extract items.
-        for blob_size in blob_sizes:
-            blob_bytes = self._crypter.decrypt(self._fp.read(blob_size))
-            offset = 0
-
-            while offset < len(blob_bytes):
-                item, offset = DatasetItemMapper.backward(blob_bytes, offset, media_path_prefix)
-                if item.media is not None and self._media_encryption:
-                    item.media.set_crypter(self._crypter)
-                self._items.append(item)
-
-            assert offset == len(blob_bytes)
+        if self._num_workers > 0:
+            self._items = self._read_items_mp(blob_sizes, media_path_prefix)
+        else:
+            self._items = self._read_items_sp(blob_sizes, media_path_prefix)
+
+        for item in self._items:
+            if item.media is not None and self._media_encryption:
+                item.media.set_crypter(self._crypter)
+
+    def _read_items_mp(
```

Comment:
Don't we need try-except statement to avoid potential faults? Is it possible to prevent all the cases by the use of timeout?

Reply:
In case of failure, we have to raise an error and notify to users that something goes wrong. The reason why I added a timeout is because the user must be notified if a situation occurs where the job never ends. If we don't have timeout, it just never stops and spins forever.

Reply:
I agree, but I have a concern about timeout. Is it okay with slow computing environment? It may require more time than our expectation to encrypt dataset. (or large dataset incomes.. with slow CPU or HDD.. )

Reply:
The timeout will be affected to each blob (not the entire dataset), but I think it sounds reasonable. Let the timeout be longer to 5 minutes. Let this value to be merged into `develop`. I'll see this value is reasonable by running manual real-life tests for a production level dataset (COCO instance segmentation train/val).

Reply:
Thank you for detailed explanation!